File: Python/ByRound/1970/1970_F1_Playing_Quidditch_Easy.py
# Problem: CF 1970 F1 - Playing Quidditch (Easy)
# https://codeforces.com/contest/1970/problem/F1

"""
Quidditch Game Scoring System

Algorithms/Techniques:
- Grid simulation with hash maps for tracking entities
- State machine for ball handling (free/carrying)
- Movement and action processing in order
- Goal detection upon ball placement

Time Complexity: O(N * M + T * K) where N and M are grid dimensions, T is number of steps, K is average moves per step (constant in practice)
Space Complexity: O(N * M + P) where P is number of players, for storing grid state and player positions

The code simulates a Quidditch match between two teams (Red and Blue), tracking player movements,
ball carrying, and scoring events. Players can move, catch the Quaffle, and throw it. When a player
throws the Quaffle into the opposing team's goal, that team scores a point. The Quaffle respawns
at the center of the field after each goal. The output lists the time and team of each goal
followed by the final score.
"""

def main():
    import sys

    data = sys.stdin.read().splitlines()
    if not data:
        return
    n, m = map(int, data[0].split())
    red_goals = set()
    blue_goals = set()
    players = {}
    quaffle = {"state": "free", "position": None}
    mid_pos = ((n - 1) // 2, (m - 1) // 2)
    index = 1
    for i in range(n):
        line = data[index].split()
        index += 1
        for j in range(m):
            cell = line[j]
            if cell == "RG":
                red_goals.add((i, j))
            elif cell == "BG":
                blue_goals.add((i, j))
            elif cell == ".Q":
                quaffle["position"] = (i, j)
            elif cell != "..":
                players[cell] = (i, j)
    T = int(data[index])
    index += 1
    moves = {"U": (-1, 0), "D": (1, 0), "L": (0, -1), "R": (0, 1)}
    red_score = 0
    blue_score = 0
    for step_index in range(T):
        line = data[index]
        index += 1
        parts = line.split()
        entity_str = parts[0]
        action = parts[1]
        ball_str = parts[2] if len(parts) >= 3 else None
        if entity_str in players:
            r0, c0 = players[entity_str]
            if action in moves:
                dr, dc = moves[action]
                new_r = r0 + dr
                new_c = c0 + dc
                players[entity_str] = (new_r, new_c)
            elif action == "C":
                if ball_str == ".Q":
                    if quaffle["state"] == "free":
                        if quaffle["position"] == (r0, c0):
                            quaffle["state"] = entity_str
            elif action == "T":
                if quaffle["state"] == entity_str:
                    quaffle["state"] = "free"
                    quaffle["position"] = (r0, c0)
        elif entity_str == ".Q":
            if quaffle["state"] == "free":
                if action in moves:
                    dr, dc = moves[action]
                    rq, cq = quaffle["position"]
                    new_r = rq + dr
                    new_c = cq + dc
                    quaffle["position"] = (new_r, new_c)
        if quaffle["state"] == "free":
            q_pos = quaffle["position"]
            if q_pos in red_goals:
                blue_score += 1
                print(f"{step_index} BLUE GOAL")
                quaffle["position"] = mid_pos
            elif q_pos in blue_goals:
                red_score += 1
                print(f"{step_index} RED GOAL")
                quaffle["position"] = mid_pos
    print(f"FINAL SCORE: {red_score} {blue_score}")


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1970/1970_F1_Playing_Quidditch_Easy.cpp
// Problem: CF 1970 F1 - Playing Quidditch (Easy)
// https://codeforces.com/contest/1970/problem/F1

/*
Purpose: This code simulates a Quidditch game between two teams (Gryffindor and Ravenclaw) on a rectangular field. 
It tracks player movements, ball carrying, and scoring events based on the rules provided.

Algorithms/Techniques:
- Grid-based simulation using 2D arrays to represent the field
- Tracking of player positions and the Quaffle's position
- State management for which player is carrying the Quaffle
- Goal detection when the Quaffle is placed in a goal

Time Complexity: O(T * N * M) where T is the number of steps, and N, M are the dimensions of the grid.
Space Complexity: O(N * M) for storing the grid and player positions.
*/

#include <stddef.h>
#include <iostream>
#include <string>
#include <utility>

using namespace std;
typedef pair<int, int> pi;
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int n, m;
  cin >> n >> m;
  pi R[10], B[10]; // Arrays to store positions of red and blue players
  string board[n][m]; // 2D grid representing the field
  pi quaff; // Position of the Quaffle
  string catcher = ""; // Tracks which player is currently carrying the Quaffle
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      string s;
      cin >> s;
      if (s == "..") {
        board[i][j] = s;
      } else if (s == ".Q") {
        quaff = {i, j}; // Initialize Quaffle position
      } else if (s == "RG" || s == "BG") {
        board[i][j] = s; // Record goal positions
      } else {
        char c = s[0];
        if (c == 'R') {
          R[s[1] - '0'] = {i, j}; // Store red player position
        } else {
          B[s[1] - '0'] = {i, j}; // Store blue player position
        }
      }
    }
  }
  int t;
  cin >> t;
  int rp = 0, bp = 0; // Scores for red and blue teams
  for (int time = 0; time < t; time++) {
    string player;
    cin >> player;
    string act;
    cin >> act;
    if (act == "C") {
      string ball;
      cin >> ball;
      catcher = player; // Mark the player as carrying the ball
    } else if (act == "T") {
      catcher = ""; // Release the Quaffle
    } else {
      // Move the player
      pi *ptr = &(player[0] == 'R' ? R : B)[player[1] - '0'];
      if (act == "U") {
        ptr->first--;
      } else if (act == "D") {
        ptr->first++;
      } else if (act == "L") {
        ptr->second--;
      } else {
        ptr->second++;
      }
    }
    // Update Quaffle position if someone is carrying it
    if (catcher != "")
      quaff = (catcher[0] == 'R' ? R : B)[catcher[1] - '0'];
    else {
      // Check if Quaffle is in a goal
      if (board[quaff.first][quaff.second] == "RG") {
        cout << time << " BLUE GOAL\n"; // Red team scores in their own goal, so Blue team gets the point
        rp++;
        quaff = {n / 2, m / 2}; // Reset Quaffle to center
      } else if (board[quaff.first][quaff.second] == "BG") {
        cout << time << " RED GOAL\n"; // Blue team scores in their own goal, so Red team gets the point
        bp++;
        quaff = {n / 2, m / 2}; // Reset Quaffle to center
      }
    }
  }
  cout << "FINAL SCORE: " << bp << " " << rp << "\n";
}


// https://github.com/VaHiX/CodeForces/